File: app/src/main/java/com/example/dices2/Player.java
package com.example.dices2;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

import static com.example.dices2.Consts.*;

public class Player implements Comparable<Player> {
    private String name;
    private Map<RowName, Integer> values;
    private boolean isSecondNonSchoolPartAdded = false;

    public Player(String name) {
        this.name = name;
        values = new HashMap<>();
    }

    public String getName() {
        return name;
    }

    public int getValue(RowName rowName) {
        return values.get(rowName);
    }

    public void setValue(RowName rowName, String value) {
        if (value.equals("")) {
            values.remove(rowName);
        } else {
            values.put(rowName, getIntValueFromString(value));
        }
    }

    public boolean isSecondNonSchoolPartAdded() {
        return isSecondNonSchoolPartAdded;
    }

    public void setSecondNonSchoolPartAdded(boolean secondNonSchoolPartAdded) {
        isSecondNonSchoolPartAdded = secondNonSchoolPartAdded;
    }

    public int getSchool() {
        int result = 0;
        for (RowName rowName : NAMES_OF_SCHOOL_ROWS) {
            Integer value = values.get(rowName);
            if (value != null) {
                result += value;
            }
        }
        if (isSchoolFinished() && result < 0) {
            result -= 50;
        }
        return result;
    }

    public int getTotal() {
        int result = 0;
        for (RowName rowName : namesOfRowsWithoutSchool) {
            Integer value = values.get(rowName);
            if (value != null) {
                result += value;
            }
        }
        return result + getSchool();
    }

    private int getIntValueFromString(String value) {
        if (value.equals("-") || value.equals("+")) {
            return 0;
        } else if (value.charAt(0) == '+') {
            return Integer.parseInt(value.substring(1));
        } else {
            return Integer.parseInt(value);
        }
    }

    public boolean isSchoolFinished() {
        for (RowName schoolClass : NAMES_OF_SCHOOL_ROWS) {
            if (values.get(schoolClass) == null) {
                return false;
            }
        }
        return true;
    }

    public boolean isThreeClassesInSchoolAreFinished() {
        int count = 0;
        for (RowName schoolClass : NAMES_OF_SCHOOL_ROWS) {
            if (values.get(schoolClass) != null && !values.get(schoolClass).equals("")) {
                count++;
            }
        }
        return count >= 3;
    }

    public void clearAllValues() {
        values.clear();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Player)) return false;
        Player player = (Player) o;
        return Objects.equals(name, player.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name);
    }

    @Override
    public int compareTo(Player player) {
        Integer temp = getTotal();
        return temp.compareTo(player.getTotal());
    }
}

File: app/src/main/java/com/example/dices2/Player.kt
package com.example.dices2

import java.util.*

class Player(val name: String) : Comparable<Player> {
    private val values: MutableMap<RowName, Int>
    var isSecondNonSchoolPartAdded = false
    var curPlace = 0

    init {
        values = HashMap()
    }

    val school: Int
        get() {
            var result = 0
            for (rowName in Consts.NAMES_OF_SCHOOL_ROWS) {
                val value = values[rowName]
                if (value != null) {
                    result += value
                }
            }
            if (isSchoolFinished && result < 0) {
                result -= 50
            }
            return result
        }

    val total: Int
        get() {
            var result = 0
            for (rowName in Consts.NAMES_OF_ROWS_WITHOUT_SCHOOL) {
                val value = values[rowName]
                if (value != null) {
                    result += value
                }
            }
            return result + school
        }

    val isSchoolFinished: Boolean
        get() {
            for (schoolClass in Consts.NAMES_OF_SCHOOL_ROWS) {
                if (values[schoolClass] == null) {
                    return false
                }
            }
            return true
        }

    val isThreeClassesInSchoolAreFinished: Boolean
        get() {
            var count = 0
            for (schoolClass in Consts.NAMES_OF_SCHOOL_ROWS) {
                if (values[schoolClass] != null) {
                    count++
                }
            }
            return count >= 3
        }

    fun getValue(rowName: RowName): Int {
        return values[rowName]?:0
    }

    fun setValue(rowName: RowName, value: String) {
        if (value == "") {
            values.remove(rowName)
        } else {
            values[rowName] = getIntValueFromString(value = value)
        }
    }

    private fun getIntValueFromString(value: String): Int {
        return if (value == "-" || value == "+") {
            0
        } else if (value[0] == '+') {
            value.substring(1).toInt()
        } else {
            value.toInt()
        }
    }

    fun clearAllValues() {
        values.clear()
    }

    fun clearPlace() {
        curPlace = 0
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Player) return false
        return name == other.name
    }

    override fun hashCode(): Int {
        return Objects.hash(name)
    }

    override fun compareTo(other: Player): Int {
        val temp = total
        return temp.compareTo(other.total)
    }
}